illumos-utils/src/contract.rs

```diff
@@ -0,0 +1,375 @@
+                // See illumos sources in `usr/src/cmd/zlogin/zlogin.c` in the
+                // implementation of `init_template()` for details.
+                if unsafe { ct_tmpl_set_critical(fd, CT_PR_EV_EMPTY) } != 0 {
+                    Err(err("set_critical in process template"))
+                } else if unsafe { ct_tmpl_set_informative(fd, 0) } != 0 {
+                    Err(err("set_informative in process template"))
+                } else if unsafe { ct_pr_tmpl_set_fatal(fd, CT_PR_EV_HWERR) }
+                    != 0
+                {
+                    Err(err("set_fatal in process template"))
+                } else if unsafe {
+                    ct_pr_tmpl_set_param(fd, CT_PR_PGRPONLY | CT_PR_REGENT)
+                } != 0
+                {
+                    Err(err("set_param in process template"))
+                } else if unsafe { ct_tmpl_activate(fd) } != 0 {
+                    Err(err("activating in process template"))
+                } else {
+                    Ok(Self { fd })
+                }
+            }
+            ContractType::Device => {
+                // The only device contract we currently support is for the
+                // tfpkt device.  If we ever want to support something else then
+                // we will need to include a device type, path, and/or instance
+                // number as an additional argument to this template creation
+                // function.
+                //
+                // Note: what we are actually interested in is the removal of
+                // the "tofino" device.  However, illumos won't report the
+                // removal of that device while it is still assigned to a zone.
+                // Since our goal is to shut down the zone to allow the removal
+                // of the device, we have a chicken/egg problem. Instead, we
+                // watch here for the removal of "tfpkt", which is a child of
+                // the "tofino", and whose removal is a reliable indicator that
+                // the tofino is gone.
+                let cpath = match get_tfpkt_device_path() {
+                    Some(c) => Ok(c),
+                    None => {
+                        unsafe { libc::close(fd) };
+                        Err(err("unable to find tfpkt in device tree"))
+                    }
+                }?;
+
+                // We want to be notified when the device goes offline
+                if unsafe { ct_tmpl_set_critical(fd, CT_DEV_EV_OFFLINE) } != 0 {
+                    Err(err("set_critical in device template"))
+                } else if unsafe { ct_dev_tmpl_set_minor(fd, cpath.as_ptr()) }
+                    != 0
+                {
+                    let cpath: Vec<u8> =
+                        cpath.iter().map(|c| *c as u8).collect();
+                    Err(err(format!(
+                        "set_minor to {} in device template",
+                        str::from_utf8(&cpath[..]).unwrap()
```

Comment:
`cpath` is coming from an external source, so it's probably best not to panic here if it somehow contains non-UTF8 data. Maybe `String::from_utf8_lossy(cpath)` instead?